### PR TITLE
Modify the implementation of KEMs and Kyber

### PR DESCRIPTION
**Pull request by dkostic**
### Issues:
CryptoAlg-1259

### Description of changes: 
- Removed the OpenSSL 3.0-like structures and APIs from the implementation of PQ KEMs.
- Added two new methods in EVP_PKEY, encapsulate and decapsulate, to be used for KEMs.
- Renamed some macro and variable names to be consistent with NIST's notation.

### Call-outs:
Point out areas that need special attention or support during the review process. Discuss architecture or design changes.

### Testing:
How is this change tested (unit tests, fuzz tests, etc.)? Are there any testing steps to be verified by the reviewer?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: crypto/kyber/kem_kyber.h [hunk 1]
```diff
@@ -6,29 +6,23 @@
...
+// pqcrystals-kyber directory.
 #define KYBER512_PUBLIC_KEY_BYTES 800
-#define KYBER512_KEM_CIPHERTEXT_BYTES 768
-#define KYBER512_KEM_SHARED_SECRET_BYTES 32
+#define KYBER512_SECRET_KEY_BYTES 1632
+#define KYBER512_CIPHERTEXT_BYTES 768
+#define KYBER512_SHARED_SECRET_BYTES 32
 
-int kyber512_keypair(uint8_t *public_key, uint8_t *secret_key);
+int kyber512_keypair(uint8_t *public_key /* OUT */,
```
[the remark this answers is not in the document]
Comment (dkostic) [reply crypto/kyber/kem_kyber.h:16]: we don't have to, I just find it useful. Would you prefer we don't deviate from the current "standard"?

File: crypto/kyber/p_kyber.c [hunk 2]
```diff
@@ -29,6 +29,56 @@ static int pkey_kyber512_keygen(EVP_PKEY_CTX *ctx, EVP_PKEY *pkey) {
...
+static int pkey_kyber512_encapsulate(EVP_PKEY_CTX *ctx,
+                                     uint8_t *ciphertext,
+                                     size_t  *ciphertext_len,
+                                     uint8_t *shared_secret,
+                                     size_t  *shared_secret_len) {
+  *ciphertext_len = KYBER512_CIPHERTEXT_BYTES;
+  *shared_secret_len = KYBER512_SHARED_SECRET_BYTES;
+  if (ciphertext == NULL) { // Caller is getting parameter values.
+    return 1;
+  }
```
Comment (andrewhop) [inline crypto/kyber/p_kyber.c:41]: If the caller is not calling to get parameters should this check that the provided buffers are big enough to hold the resulting cipher text and shared secret before updating the sizes?
```suggestion
  if (ciphertext == NULL) { // Caller is getting parameter values.
    *ciphertext_len = KYBER512_CIPHERTEXT_BYTES;
    *shared_secret_len = KYBER512_SHARED_SECRET_BYTES;
    return 1;
  } else if (*ciphertext_len < KYBER512_CIPHERTEXT_BYTES || shared_secret){
    //error
  } else {
    *ciphertext_len = KYBER512_CIPHERTEXT_BYTES;
    *shared_secret
  }
```

File: include/openssl/evp.h [hunk 3]
```diff
@@ -690,20 +690,49 @@ OPENSSL_EXPORT int EVP_PKEY_keygen_init(EVP_PKEY_CTX *ctx);
...
+//      writes it to |shared_secret|,
+//   3. encapsulates the shared secret, producing the ciphertext, by using
+//      the public key in |ctx| and writes the ciphertext to |ciphertext|.
+//
+// If the given |ciphertext| is NULL it is assumed that the caller is doing
+// a size check: the function will write the size of the ciphertext and the
+// shared secret in |ciphertext_len| and |shared_secret_len| and return 1.
+//
+// NOTE: no allocation is done in the function, the caller is expected to
+// provide large enough |ciphertext| and |shared_secret| buffers.
```
Comment (andrewhop) [inline include/openssl/evp.h:707]: Do other EVP functions check/enforce this or just rely on customers doing the right thing and testing with valgrind/address sanitizer?

Do we have tests for bad input sizes? Do we have a SIM for fuzz tests?

File: crypto/evp_extra/p_hkdf.c [hunk 4]
```diff
@@ -207,6 +207,8 @@ const EVP_PKEY_METHOD hkdf_pkey_meth = {
     pkey_hkdf_derive,
     /*paramgen=*/NULL,
     pkey_hkdf_ctrl,
+    /*encapsulate*/NULL,
```
Comment (andrewhop) [inline crypto/evp_extra/p_hkdf.c:210]: This order of comment then null or null then comment is silly. I see that you're just being consistent

File: crypto/kyber/p_kyber.c (hunk 2 above)
Comment (dkostic) [reply crypto/kyber/p_kyber.c:41]: That's a good question, I'm not sure how was this part of the API designed. I assumed `ciphertext_len` and `shared_secret_len` are output arguments so we can't perform checks on them.

Comment (dkostic) [reply crypto/kyber/p_kyber.c:41]: Actually, you are right, I'm reading the documentation of the other functions in EVP_PKEY and they all expect `*_len` to have the length of the output buffer. I'll address this.

Comment (dkostic) [reply crypto/kyber/p_kyber.c]: addressed in the latest change.

File: include/openssl/evp.h (hunk 3 above)
Comment (dkostic) [reply include/openssl/evp.h]: > Do other EVP functions check/enforce this or just rely on customers doing the right thing and testing with valgrind/address sanitizer?

I haven't gone through all the EVP function, but the ones I did go through check if the `*_len` value is large enough. Not sure what you mean by "enforcing"?

> Do we have tests for bad input sizes? Do we have a SIM for fuzz tests?

I added a test for too small input sizes in `p_kyber_test.cc`. The SIM for fuzz tests: CryptoAlg-1295.

File: crypto/kyber/kem_kyber.h [hunk 5]
```diff
@@ -6,29 +6,23 @@
 #include <openssl/base.h>
 #include <openssl/evp.h>
 
-#define KYBER512_PRIVATE_KEY_BYTES 1632
+// The values below are taken from the |api.h| file in the
+// pqcrystals-kyber directory.
```
[the remark this answers is not in the document]
Comment (torben-hansen) [reply crypto/kyber/kem_kyber.h]: Okay, we have to keep them in sync now?

File: crypto/kyber/p_kyber512_asn1.c [hunk 6]
```diff
@@ -16,7 +16,7 @@ static void kyber512_free(EVP_PKEY *pkey) {
 }
 
 static int kyber512_set_priv_raw(EVP_PKEY *pkey, const uint8_t *in, size_t len) {
-  if (len != KYBER512_PRIVATE_KEY_BYTES) {
+  if (len != KYBER512_SECRET_KEY_BYTES) {
```
Comment (torben-hansen) [inline crypto/kyber/p_kyber512_asn1.c:19]: Can we statically assert there is space enough in `key->priv` in line 30? (https://github.com/awslabs/aws-lc/pull/605/files#diff-283a395b51f7e614369b99e205f13b02151037f16d48f4e5ca43733c93fd3554R30)

Comment (torben-hansen) [reply crypto/kyber/p_kyber512_asn1.c:19]: Same for the public field in `kyber512_set_pub_raw()`.

File: crypto/kyber/p_kyber512_asn1.c [hunk 7]
```diff
@@ -103,11 +103,11 @@ static int kyber512_pub_cmp(const EVP_PKEY *a, const EVP_PKEY *b) {
 }
 
 static int kyber512_size(const EVP_PKEY *pkey) {
-  return KYBER512_PUBLIC_KEY_BYTES + KYBER512_PRIVATE_KEY_BYTES;
+  return KYBER512_PUBLIC_KEY_BYTES + KYBER512_SECRET_KEY_BYTES;
 }
 
 static int kyber512_bits(const EVP_PKEY *pkey) {
-  return 8 * (KYBER512_PUBLIC_KEY_BYTES + KYBER512_PRIVATE_KEY_BYTES);
+  return 8 * (KYBER512_PUBLIC_KEY_BYTES + KYBER512_SECRET_KEY_BYTES);
```
Comment (torben-hansen) [inline crypto/kyber/p_kyber512_asn1.c:110]: huge nit, but still:
```suggestion
  return 8 * kyber512_size();
```

Comment (torben-hansen) [reply crypto/kyber/p_kyber512_asn1.c:110]: Also, the documentation for the public function that uses this call-back is not very precise on what the expected value is for KEMs:
```
// EVP_PKEY_size returns the maximum size, in bytes, of a signature signed by
// |pkey|. For an RSA key, this returns the number of bytes needed to represent
// the modulus. For an EC key, this returns the maximum size of a DER-encoded
// ECDSA signature.
OPENSSL_EXPORT int EVP_PKEY_size(const EVP_PKEY *pkey);
```
Hopefully, the returned value is not going to change. But we'll see whether the final standard is going to introduce some encoding...

File: crypto/kyber/p_kyber512_asn1.c [hunk 8]
```diff
@@ -103,11 +103,11 @@ static int kyber512_pub_cmp(const EVP_PKEY *a, const EVP_PKEY *b) {
 }
 
 static int kyber512_size(const EVP_PKEY *pkey) {
```
Comment (torben-hansen) [inline crypto/kyber/p_kyber512_asn1.c:105]: Are these functions exercised in a test function anywhere?

File: crypto/kyber/p_kyber_test.cc [hunk 9]
```diff
@@ -215,39 +213,54 @@ TEST(Kyber512Test, KEMSizeChecks) {
...
+  EXPECT_EQ(shared_secret_len, (size_t)KYBER512_SHARED_SECRET_BYTES);
+  EXPECT_EQ(ciphertext_len, (size_t)KYBER512_CIPHERTEXT_BYTES);
 
   shared_secret_len = 0;
 
   ASSERT_TRUE(EVP_PKEY_decapsulate(kyber_pkey_ctx, NULL, &shared_secret_len, NULL, ciphertext_len));
-  EXPECT_EQ(shared_secret_len, (size_t)KYBER512_KEM_SHARED_SECRET_BYTES);
+  EXPECT_EQ(shared_secret_len, (size_t)KYBER512_SHARED_SECRET_BYTES);
+
+  // Verify that encaps/decaps fail when given too small buffer lengths.
```
Comment (torben-hansen) [inline crypto/kyber/p_kyber_test.cc:225]: Nice

File: include/openssl/evp.h [hunk 10]
```diff
@@ -690,20 +690,53 @@ OPENSSL_EXPORT int EVP_PKEY_keygen_init(EVP_PKEY_CTX *ctx);
...
+OPENSSL_EXPORT int EVP_PKEY_encapsulate(EVP_PKEY_CTX *ctx          /* IN  */,
+                                        uint8_t *ciphertext        /* OUT */,
+                                        size_t  *ciphertext_len    /* OUT */,
+                                        uint8_t *shared_secret     /* OUT */,
+                                        size_t  *shared_secret_len /* OUT */);
+
+// EVP_PKEY_decapsulate is an operation defined for a KEM (Key Encapsulation
+// Mechanism). For the KEM specified in |ctx|, the function:
+//   1. decapsulates the shared secret from the given |ciphertext| using the
+//      secret key given within |ctx| and writes it to |shared_secret|,
```
Comment (torben-hansen) [inline include/openssl/evp.h:721]: Just a nit suggestion:
```suggestion
//      secret key configured in |ctx| and writes it to |shared_secret|,
```

File: crypto/kyber/p_kyber512_asn1.c (hunk 6 above)
Comment (dkostic) [reply crypto/kyber/p_kyber512_asn1.c:19]: done

File: crypto/kyber/p_kyber512_asn1.c (hunk 7 above)
Comment (dkostic) [reply crypto/kyber/p_kyber512_asn1.c]: added a sentence for KEMs

File: crypto/kyber/p_kyber512_asn1.c (hunk 8 above)
Comment (dkostic) [reply crypto/kyber/p_kyber512_asn1.c:111]: yes, in `crypto/kyber/p_kyber_test.cc`

**Review by torben-hansen — DISMISSED** [review]

File: crypto/fipsmodule/evp/internal.h [hunk 11]
```diff
@@ -255,6 +244,16 @@ struct evp_pkey_method_st {
...
+
+  // Encapsulate and decapsulate are operations defined for a
+  // Key Encapsulation Mechanism (KEM).
+  int (*encapsulate)(EVP_PKEY_CTX *ctx,
+                     uint8_t *ciphertext, size_t *ciphertext_len,
+                     uint8_t *shared_secret, size_t *shared_secret_len);
+
+  int (*decapsulate)(EVP_PKEY_CTX *ctx,
+                     uint8_t *shared_secret, size_t *shared_secret_len,
+                     uint8_t *ciphertext, size_t ciphertext_len);
```
[the remark this answers is not in the document]
Comment (dkostic) [reply crypto/fipsmodule/evp/internal.h]: done

File: crypto/kyber/kem_kyber512.c [hunk 12]
```diff
@@ -0,0 +1,27 @@
+#include "../evp_extra/internal.h"
+#include "../fipsmodule/evp/internal.h"
+#include "kem_kyber.h"
+#include "pqcrystals-kyber_kyber512_ref/api.h"
```
[the remark this answers is not in the document]
Comment (dkostic) [reply crypto/kyber/kem_kyber512.c:4]: done

File: crypto/kyber/p_kyber.c [hunk 13]
```diff
@@ -29,6 +29,80 @@ static int pkey_kyber512_keygen(EVP_PKEY_CTX *ctx, EVP_PKEY *pkey) {
...
+  if (ciphertext == NULL) {
+    *ciphertext_len = KYBER512_CIPHERTEXT_BYTES;
+    *shared_secret_len = KYBER512_SHARED_SECRET_BYTES;
+    return 1;
+  }
+
+  // The output buffers need to be large enough.
+  if (*ciphertext_len < KYBER512_CIPHERTEXT_BYTES ||
+      *shared_secret_len < KYBER512_SHARED_SECRET_BYTES) {
+      return 0;
```
[the remark this answers is not in the document]
Comment (dkostic) [reply crypto/kyber/p_kyber.c:48]: done

**Review by torben-hansen — APPROVED** [review]

File: crypto/kyber/p_kyber.c [hunk 14]
```diff
@@ -29,6 +29,80 @@ static int pkey_kyber512_keygen(EVP_PKEY_CTX *ctx, EVP_PKEY *pkey) {
...
+  *shared_secret_len = KYBER512_SHARED_SECRET_BYTES;
+
+  return 1;
+}
+
+static int pkey_kyber512_decapsulate(EVP_PKEY_CTX *ctx,
+                                     uint8_t *shared_secret,
+                                     size_t  *shared_secret_len,
+                                     uint8_t *ciphertext,
+                                     size_t   ciphertext_len) {
```
[the remark this answers is not in the document]
Comment (dkostic) [reply crypto/kyber/p_kyber.c:75]: added a check for `ciphertext_len` inside the function so now it's used :)

**Review by andrewhop — APPROVED** [review]

**Review by torben-hansen — APPROVED** [review]

**Pull request merged**